oximeter/db/tests/integration_test.rs

```diff
@@ -0,0 +1,380 @@
+
+    // Wait for all the data to be copied to node 3
+    let start = tokio::time::Instant::now();
+    wait_for_num_points(&client3, samples.len())
+        .await
+        .expect("failed to get samples from client3");
+    let end = tokio::time::Instant::now();
+    println!("query samples from client3 time = {:?}", end - start);
+
+    // Let's stop replica 1 and write some data to replica 3
+    // When we bring replica 1 back up we should see data get replicated to it.
+    // Data should also get replicated to replica 2 immediately.
+    deployment.stop_server(1).unwrap();
+
+    // Generate some new samples and insert them at replica3
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    client3.insert_samples(&samples).await.expect("failed to insert samples");
+
+    // Ensure that both replica 3 and replica 2 have the samples
+    wait_for_num_points(&client3, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client3");
+    wait_for_num_points(&client2, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client2");
+
+    // Restart node 1 and wait for the data to replicate
+    deployment.start_server(1).expect("failed to restart clickhouse server 1");
+    wait_for_ping(&client1).await.expect("failed to ping server 1");
+    wait_for_num_points(&client1, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client1");
+
+    // Remove a keeper node
+    deployment.remove_keeper(2).expect("failed to remove keeper");
+
+    // Querying from any node should still work after a keeper is removed
+    wait_for_num_points(&client1, samples.len() * 2)
+        .await
+        .expect("failed to get samples from client1");
+
+    // We still have a quorum (2 of 3 keepers), so we should be able to insert
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    client3.insert_samples(&samples).await.expect("failed to insert samples");
+    wait_for_num_points(&client2, samples.len() * 3)
+        .await
+        .expect("failed to get samples from client1");
+
+    // Stop another keeper
+    deployment.stop_keeper(1).expect("failed to stop keeper");
+
+    // We should still be able to query
+    wait_for_num_points(&client3, samples.len() * 3)
+        .await
+        .expect("failed to get samples from client1");
+
+    println!("Attempting to insert samples without keeper quorum");
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    // We have lost quorum and should not be able to insert
+    client1
+        .insert_samples(&samples)
+        .await
+        .expect_err("insert succeeded without keeper quorum");
```

Comment:
Can you describe how this error manifests? Is it immediate? A timeout?

Reply:
Unfortunately this manifests as a timeout. I will add a comment about this.

Double unfortunate is that it requires a relatively large timeout on the client side given the amount of time taken to initialize the database and the fact that we are sharing a single client here. Maybe I should create a new client with a smaller timeout to minimize the test run time.

Reply:
Done in [29bd934](https://github.com/oxidecomputer/omicron/pull/6131/commits/29bd9341c9f5a7da42d5482ff4a162f80e929309)